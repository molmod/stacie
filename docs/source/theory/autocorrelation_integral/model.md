# Model Spectrum

Stacie supports two models for fitting the low-frequency part of the power spectrum.
In both models, the value at zero frequency corresponds to the autocorrelation integral.

1. The [ExpPolyModel](#stacie.model.ExpPolyModel) is the most general:
   it is an exponential function of a linear combination of simple monomials of the frequency.
   One can specify the degrees of the monomials and typically a low degree works fine:

    - Degree `[0]` is suitable for a white noise spectrum.
    - Degrees `[0, 1]` can be used to extract useful information of a very noisy spectrum.
    - Degrees `[0, 1, 2]` is applicable to spectra with low statistical uncertainty,
      e.g. averaged over 100 inputs.
    - An even polynomial with degrees `[0, 2]` is suitable for spectra with a peak at zero frequency.

    The main advantage of this model is its broad applicability,
    as it requires little prior knowledge of the functional form of the spectra.

2. The Exponential Tail model is designed for autocorrelation functions that decay exponentially.
   Its primary advantage is that, in addition to the integrated correlation time,
   it also provides an estimate of the exponential correlation time.

## 1. ExpPolyModel

The [ExpPolyModel](#stacie.model.ExpPolyModel) is defined as

$$
    C^\text{exppoly}_k = \exp\left(\sum_{s\in S} a_s f_k^s\right)
$$

where $S$ is the set of polynomial degrees.
With this form, $a_0$ corresponds to the integral of the autocorrelation function.
When one obtains an estimate $\hat{a}_0$ and an estimated variance $\hat{C}(a_0)$,
the autocorrelation integral is [log-normally distributed](https://en.wikipedia.org/wiki/Log-normal_distribution)
with estimated mean and variance:

$$
    \hat{\mathcal{I}}
    &= \exp\left(\hat{a}_0 + \frac{1}{2}\hat{C}(a_0)\right)
    \\
    \hat{C}(\mathcal{I})
    &= \exp\left(2\hat{a}_0 + \hat{C}(a_0)\right)
    \left(\exp(\hat{C}(a_0)) - 1 \right)
$$

## 2. Exponential Tail Model

This model represents the autocorrelation function
as the sum of a short-term component and a (periodic) exponentially decaying tail:

$$
c_\Delta = c_\Delta^\text{short} + c_\Delta^\text{tail}
$$

### Short-term component

The short-term component is non-zero only for small positive or negative time lags:

$$
c_\Delta^\text{short} \neq 0 & \quad\text{if } |\Delta| \le \Delta_\text{short}
\\
c_\Delta^\text{short} = 0 & \quad\text{if } \Delta_\text{short} \lt |\Delta| \le N/2
$$

Because summations in DFT are commonly taken from $0$ to $N-1$, we rewrite this as:

$$
c_\Delta^\text{short}
    \neq 0 &\qquad \forall\,\Delta\in\lbrace
        0, \ldots, \Delta_\text{short},
        N-\Delta_\text{short}, \ldots, N-1
    \rbrace
\\
c_\Delta^\text{short}
    = 0 &\qquad \forall\,\Delta\in\lbrace
        \Delta_\text{short}+1, \ldots, N-\Delta_\text{short} -1
    \rbrace
$$

### Tail Component

The tail component is the periodic repetition (sum over $i$) of two exponential functions:
one decaying for positive time lags and one for negative time lags.
For $\Delta \in \lbrace 0, \ldots, N-1 \rbrace$, this can be written as:

$$
c_\Delta^\text{tail}
    = \frac{a_\text{tail}}{M} (r^\Delta + r^{N-\Delta}) \sum_{i=0}^\infty r^{iN}
$$

with $r < 1$.
The factor $\frac{1}{M}$ normalizes the tail component so that it sums to $a_\text{tail}$.

$$
M
    &= \left(\frac{1-r^N}{1-r} + r^N \frac{1-r^{-N}}{1-r^{-1}}\right) \sum_{i=0}^\infty r^{iN}
\\
    &= \left(1-r^{-N}\right) \frac{1 + r}{1-r} \sum_{i=0}^\infty r^{iN}
$$

We may absorb the infinite sum into the normalization constant to simplify the model:

$$
c_\Delta^\text{tail}
    & = \frac{a_\text{tail}}{M'} (r^\Delta + r^{N-\Delta})
\\
M'
    &= \left(1-r^{-N}\right) \frac{1 + r}{1-r}
$$

The exponential decay of the tail component is characterized by
its autocorrelation time, $\tau_\text{exp}$ {cite:p}`sokal_1997_monte`:

$$
    r = \exp\left(-\frac{h}{\tau_\text{exp}}\right)
$$

where $h$ is the time step.

### Discrete Fourier Transform

For small values of $k$, the discrete Fourier transform of the model autocorrelation function becomes:

$$
C^\text{exp-tail}_k
    &\approx
    a_\text{short} + \frac{a_\text{tail}}{M'} \left(
        \frac{1-r^N}{1 - r \omega^{-k}} + r^N\frac{1-r^{-N}}{1 - r^{-1} \omega^{-k}}
    \right)
\\
    &\approx
    a_\text{short} + \frac{a_\text{tail}}{M'} \left(
        \frac{1-r^N}{1 - r \omega^{-k}} + \frac{1-r^N}{1 - r \omega^k} - \left(1-r^N\right)
    \right)
\\
    &\approx
    a_\text{short} + \frac{a_\text{tail}}{M'} \left(1-r^N\right) \left(
         \frac{2 - r(\omega^k + \omega^{-k})}{1 - r(\omega^k + \omega^{-k}) + r^2} - 1
    \right)
\\
    &\approx
    a_\text{short} + a_\text{tail} \frac{1-r}{1+r} \left(
         \frac{2 - r(\omega^k + \omega^{-k})}{1 - r(\omega^k + \omega^{-k}) + r^2} - 1
    \right)
$$

For the first term we assumed $\omega^{k\Delta_\text{short}}\approx 1$.
Finally, we can substitute $\omega^k + \omega^{-k} = 2\cos(2\pi k/N)$:

$$
C^\text{exp-tail}_k
    &\approx
    a_\text{short} + a_\text{tail} \frac{1-r}{1+r} \left(
         2\frac{1 - r\cos(2\pi k/N)}{1 - 2r\cos(2\pi k/N) + r^2} - 1
    \right)
    \\
    &\approx
    a_\text{short} + a_\text{tail} \frac{(1-r)^2}{1 - 2r\cos(2\pi k/N) + r^2}
    \\
$$

This model can be fitted to the low-frequency part of an empirical autocorrelation function.
Once the parameters are fitted to an empirical spectrum, one finds:

$$
    \mathcal{I} \approx \hat{\mathcal{I}} = \hat{C}^\text{exp-tail}_0 = \hat{a}_\text{short} + \hat{a}_\text{tail}
$$

### Peak width

As illustrated in the
[Exponential Tail Model](../../examples/model.py)
example, the exponential decay of the autocorrelation function results in
a peak in the power spectrum at zero frequency.
The width of this peak at half the maximum of the tail term is found by solving:

$$
    1 - 2 r \cos(2\pi k_\text{half}/N) + r^2 = 2 (1 - r)^2
$$

The solution is:

$$
    k_\text{half}
        &= \frac{N}{2\pi}\arccos\left(\frac{4r - r^2 -1 }{2r}\right)
    \\
        &= \frac{N}{2\pi}\arccos\left(2 - \cosh\left(\frac{h}{\tau_\text{exp}}\right)\right)
$$

Using RFFT conventions, $f=k/hN$, this can be rewritten as a frequency:

$$
    f_\text{half}
        = \frac{1}{2\pi h}\arccos\left(2 - \cosh\left(\frac{h}{\tau_\text{exp}}\right)\right)
$$

We can rely on the following series expansion:

$$
    \arccos(2 - \cosh(u))
    \approx
    \left|u + \frac{1}{12} u^3 + \frac{1}{96} u^5 + \mathcal{O}(u^7)\right|
$$

Hence, in the limit $\tau_\text{exp} \gg h$, we find:

$$
    f_\text{half}
        \approx \frac{1}{2\pi \tau_\text{exp}}
$$

Sequences obtained from computer simulations
usually have sufficiently small time steps to satisfy this limit:
The steps must be able to resolve the fastest oscillations in the system
to correctly simulate its dynamics.
