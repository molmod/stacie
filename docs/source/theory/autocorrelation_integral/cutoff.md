# Frequency Cutoff

The Exponential Tail model should be fitted to the low-frequency part of the spectrum,
but it is not clear *a priori* up to which frequency.
Stacie's strategy is to test several frequency cutoffs, uniformly distributed on a log scale.
The best cutoff is identified as the one
that maximizes the amount of data included without an obvious trend in the residuals.
Including more data points reduces the variance of the autocorrelation integral estimate,
but any trend in the residuals would be a sign of underfitting, leading to biased parameters.

Visually recognizing underfitting may seem trivial,
but designing a robust criterion to quantify it is not.
Traditional model selection criteria such as
[Marginal Likelihood](https://en.wikipedia.org/wiki/Marginal_likelihood),
[Akaike's Information Criterion (AIC)](https://en.wikipedia.org/wiki/Akaike_information_criterion),
or [Bayesian Information Criterion](https://en.wikipedia.org/wiki/Bayesian_information_criterion)
are not trivially applicable because they solve a different problem:
identifying the best model for a given data set.
Here, the model is the same for all cutoffs, but the amount of fitting data varies.

A new "underfitting criterion" (UFC) for selecting the best cutoff frequency is presented below.
Its value decreases slowly as more data is included and as long as the residuals exhibit no trends.
As soon as any trend appears in the residuals, the underfitting criterion increases sharply.
The amount of data for which the criterion becomes minimal represents a sweet spot:
the amount of useful information is maximized,
subject to the constraint that there is no risk of biasing the parameters.

The UFC and AIC (with a workaround discussed below) can be used to select a domain
when fitting a smooth univariate model to noisy data.
Both criteria are illustrated for a simple polynomial fit to noisy cosine data
in the [Underfitting Criterion Demo](../../examples/underfitting_criterion.py).
This example may be helpful to gain some intuition for the theory derived below.

## Normalized Residuals

The underfitting criterion is applicable to any 1D curve fitting problem
where only a part of the data is expected to follow the model.
This "part" is not known *a priori*,
and an appropriate range of the dependent variables is searched.
For each selected range, the model is fitted to $N_\text{fit}$ data points.
The corresponding data points are labeled with an integer index $i \in \{1 \ldots N_\text{fit}\}$.
The index must preserve the order of the independent variables.
For example, in the case of a spectrum, from low frequency to high frequency.

The underfitting criterion uses normalized residuals after maximizing the likelihood:

$$
    \hat{R}_i = \frac{\hat{Y}_i - \hat{\mu}_i}{\hat{\sigma}_i}
$$

where $\hat{\mu}_i$ and $\hat{\sigma}_i$
are the maximum-likelihood mean and standard deviation of the model for data point $i$.
In the case of spectra, the residuals in the notation from the
[Parameter Estimation](../model.md#lmax-target) section are:

$$
  \hat{Y}_i &= \hat{C}_i
  \\
  \hat{\mu}_i &= \kappa_i\,\hat{\theta}_i
  \\
  \hat{\sigma}_i &= \sqrt{\kappa_i}\,\hat{\theta}_i
$$

The parameter $\hat{\theta}_i$ is the scale of the Gamma distribution at each frequency,
found with the best parameters, i.e. those that maximize the log-likelihood.

In a regression problem (without underfitting), we would have the following expectation values:

$$
    \mean[\hat{R}_i] = 0 \qquad \mean[\hat{R}^2_i] < 1
$$

Although the residuals have been normalized, their variance should be less than one.
On average,
the maximum likelihood model is (slightly) closer to the measured data than to the ground truth
because it tries to make the residuals as small as possible.
In other words, on average, likelihood maximization will result in a slightly overfitted model.

If a model is too simple to describe the data,
there will be some underfitting and the normalized residuals can become (much) larger than one.
Therefore, in principle, a sample variance of the normalized residuals
greater than one can be used to detect underfitting.
However, such a simple criterion does not detect small misfits.
In edge cases, the variance of the normalized residuals may still be close to one,
even though they show a trend.

## Cumulative Sums of Residuals

Underfitting can be detected more robustly by
making use of correlations between nearby residuals.
For example, the analytical spectrum of the Exponential Tail model is smooth
compared to the uncorrelated uncertainties of a power spectrum.
If the model cannot fully describe the data,
nearby residuals will have the same sign on average.

One can amplify correlations between nearby residuals
by using a shifted cumulative sum of residuals:

$$
    \hat{U}_j
        = 2\sum_{i=1}^j \hat{R}_i - \sum_{i=1}^{N_\text{fit}} \hat{R}_i
           \qquad \forall\,j \in \lbrace 0, \ldots, N_\text{fit}\rbrace
$$

The first term is twice the cumulative sum of the residuals,
and the second term shifts them down by the sum of all residuals.
This shifted cumulative sum can be rewritten as:

$$
    \hat{U}_j
        = \sum_{i=1}^j \hat{R}_i - \sum_{i=j+1}^{N_\text{fit}} \hat{R}_i
$$

In this form, $\hat{U}_j$ is the sum of all residuals up to index $j$,
minus the sum of the remaining residuals.
If $j$ corresponds to the data point where the model goes from over- to undershooting the data,
then $\hat{U}_j$ will add systematic errors on both sides of $j$ with the same sign.

In the absence of underfitting, the corresponding expectation values should satisfy:

$$
    \mean[\hat{U}_j] = 0 \qquad \mean[\hat{U}^2_j] < N_\text{fit}
$$

However, in case of underfitting, there is a trend in the residuals,
meaning that nearby residuals are positively correlated.
In this case, the positively correlated errors will add up in the cumulative sums,
such that the sample variance of $\hat{U}^2_j$ quickly exceeds the expectation value above.

## Underfitting Criterion

We propose the following criterion to detect underfitting artifacts:

$$
    \operatorname{UFC}
        = \frac{1}{N_\text{fit} + 1}
          \left(\sum_{j=0}^{N_\text{fit}} \hat{U}^2_j\right)
          - N_\text{fit}
$$

Averaging over all indexes $j$ in the first term has the advantage that the UFC accounts for
all possible indexes $j$ where systematic errors may change sign.

If there is no underfitting, the UFC will be negative.
Empirically, we observe that the UFC decreases slowly with increasing amounts of data,
e.g., increasing frequency cutoff,
as long as there is no underfitting.
As soon as the model begins to underfit the data,
the sample variance of the cumulative sums and the UFC increase rapidly.
The frequency cutoff for which the UFC is minimal
always seems to strike a good balance between over- and underfitting.
The minimum is located at the onset of underfitting,
which is barely visible in the residuals to the naked eye.
At this point, one has (nearly) the maximum amount of data in the fit
(and thus minimal uncertainty of the parameters)
without biasing the estimated parameters due to underfitting.

## Akaike's Information Criterion

Akaike's Information Criterion (AIC) is intended for model selection, not data selection.
At first glance, it does not seem applicable to frequency cutoff selection.
However, it is possible reformulate data selection as model selection.

The "trick" is to introduce an additional dummy parameter to "fit" the discarded data point.
By introducing such dummy parameters,
all data in the spectrum is always included in the likelihood (regardless of the cutoff).
The cutoff simply determines at which part is modeled
with the Exponential Tail model or the dummy parameters.
Finding the optimal cutoff is thus reduced to a model selection problem.

In the case of the frequency cutoff,
the dummy parameter corresponds to the scale parameter of the Gamma distribution.
Assuming that $\kappa_k > 1$, the probability for this point is maximized by:

$$
    \hat{\theta}_k = \frac{\hat{C}_k}{\kappa_k - 1}.
$$

The log likelihood for such a point becomes:

$$
    \ln \mathcal{L}_k^\text{cut} =
        -\ln\Gamma(\kappa_k)
        - \ln\left(\frac{\hat{C}_k}{\kappa_k - 1}\right)
        + (\kappa_k - 1) \Bigl(\ln(\kappa_k - 1) - 1\Bigr)
$$

To obtain the total log likelihood, $\ln\mathcal{L}^\text{total}$,
the terms $\ln \mathcal{L}_k^\text{cut}$ can be summed over all discarded points
and added to the likelihood of the model fitted to the data before the cutoff.
The AIC with this likelihood then becomes:

$$
    \operatorname{AIC}
    = 2 (N_\text{par} + N_\text{discard}) - 2 \ln\mathcal{L}^\text{total}
$$

where:

- $N_\text{par}$ is the number of parameters in the model for the low-frequency part of the spectrum
  (3 in case of the Exponential Tail model),
- $N_\text{discard}$ is the number of discarded data points

## Cutoff selection

The AIC is the more fundamental of the two criteria and it is also widely used.
Nevertheless, in our numerical benchmarks,
the frequency cutoff that minimizes the AIC always leads to visually clearly biased fits.
In contrast, a cutoff that minimizes the UFC always leads to residuals without a noticeable trend.

We must emphasize that the UFC is an ad hoc construction.
It is nothing more than a robust mathematical implementation
of how one would intuitively describe the underfitting of a smooth 1D curve to noisy data.
Nevertheless, its use is well justified by its practical utility,
its simplicity, its effectiveness in our benchmarks,
and the complete absence of manually tunable settings.
