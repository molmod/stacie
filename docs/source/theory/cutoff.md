# Frequency Cutoff

The Exponential Tail model should be fitted to the low-frequency part of the spectrum,
but it is not clear *a priori* up to which frequency.
Stacie's strategy is to test several frequency cutoffs, uniformly distributed on a log scale.
The best cutoff is identified as the one
that maximizes the amount of included data without an apparent trend the residuals.
Any trend in the residuals would be a sign of underfitting, leading to biased parameters.

Visually recognizing underfitting may seem trivial,
but designing a robust criterion to quantify it is not.
Traditional model selection criteria,
such as [Marginal Likelihood](https://en.wikipedia.org/wiki/Marginal_likelihood),
[Akaike's Information Criterion (AIC)](https://en.wikipedia.org/wiki/Akaike_information_criterion),
or [Bayesian Information Criterion](https://en.wikipedia.org/wiki/Bayesian_information_criterion)
are not trivially applicable because they solve a different problem:
they identify the best model for a given data set.
Here, the model is the same for all cutoffs, but the amount of fitting data varies.

A new "underfitting criterion" (UFC) for selecting the best cutoff frequency is presented below.
Its value slowly decreases as more data is included and as long as the residuals exhibit no trends.
As soon as any trend appears in the residuals, the underfitting criterion sharply increases.
The amount of data for which the criterion becomes minimal represents a sweet spot:
the amount of useful information is maximized,
subject to the constraint that there is no risk of biasing the parameters.

Both AIC and UFC can be used for the selection of a domain
when fitting a smooth univariate model to noisy data.
They is illustrated for a simple polynomial fit to noisy cosine data
in the [Underfitting Criterion Demo](../examples/underfitting_criterion.py).
This example may be helpful to gain some intuition for the theory derived below.


## Normalized Residuals

The underfitting criterion, uses normalized residuals after maximization of the likelihood:

$$
    \hat{R}_k = \frac{\hat{Y}_k - \hat{\mu}_k}{\hat{\sigma}_k}
$$

where $\hat{\mu}_k$ and $\hat{\sigma}_k$ are the maximum-likelihood mean and standard deviation of the model for data point $k$. In the case of spectra, these required quantities, using the from the [Parameter Estimation](#lmax-target) section, are:

$$
  \hat{Y}_k &= \hat{C}_k
  \\
  \hat{\mu}_k &= \kappa_k\,\hat{\theta}_k
  \\
  \hat{\sigma}_k &= \sqrt{\kappa_k}\,\hat{\theta}_k
$$

The parameter $\hat{\theta}_k$ is the scale of the Gamma distribution at each frequency,
found with the best parameters, i.e. those that maximize the log-likelihood.

In the remainder of this section, we will relabel the indices $k$ of the residuals
with an integer $i$ going from $1$ to $N_\text{res}$
maintaining the same order as the index $k$ (from low to high frequency).

In a regression problem (without underfitting), we would have the following expectation values:

$$
    \mean[\hat{R}_i] = 0 \qquad \mean[\hat{R}^2_i] < 1
$$

Despite having normalized the residuals, their variance should be less than one.
On average, the maximum likelihood model will be (slightly) closer to the measurement data
than the ground truth, because it tries to make the residuals as small as possible.

When a model is too simple to describe the data,
there will be some underfitting and the normalized residuals can become (much) larger than 1.
Therefore, in principle, a sampling variance of the normalized residuals
greater than 1 can be used to detect underfittng.
However, such a simple criterion does not easily reveal slight misfits.
In edge cases, the variance of the normalized residuals can still be about 1,
despite the fact that they exhibit a trend.
Such a simple criterion is therefore not suitable for selecting an appropriate cutoff.


## Cumulative Sums of Residuals

Underfitting can be detected more robustly by
making use of correlations between nearby residuals.
The analytical spectrum of the Exponential Tail model is smooth
compared to the uncorrelated uncertainties of a power spectrum.
When the model cannot fully describe spectral data,
nearby residuals will, on average, have the same sign.

One can amplify correlations between nearby residuals
by using a shifted cumulative sum of residuals:

$$
    \hat{U}_j
        &= 2\sum_{i=1}^j \hat{R}_i - \sum_{i=1}^{N_\text{res}} \hat{R}_i
           \qquad \forall\,j \in \lbrace 0, \ldots, N_\text{res}\rbrace
    \\
        &= \sum_{i=1}^j \hat{R}_i - \sum_{i=j+1}^{N_\text{res}} \hat{R}_i
$$

$\hat{U}_j$ can also be interpreted as the sum of all residuals up to index $j$,
minus the sum of the remaining residuals.
If $j$ corresponds to the data point where the model goes from over- to undershooting the data,
then the corresponding $\hat{U}_j$ will add systematic errors on both sides of $j$ with the same sign.

In the absence of underfitting, the corresponding expectation values should satisfy:

$$
    \mean[\hat{U}_j] = 0 \qquad \mean[\hat{U}^2_j] < N_\text{res}
$$

However, in case of underfitting, there is a trend in the residuals,
meaning that nearby residuals are positively correlated.
In this case, the positively correlated errors will add up in the cumulative sums,
such that the empirical variance of $\hat{U}^2_j$ quickly exceeds the expectation value above.


## Underfitting Criterion

We propose the following criterion to detect underfitting artifacts:

$$
    \operatorname{UFC}
        = \frac{1}{N_\text{res} + 1} \sum_{j=0}^{N_\text{res}}
           \left(\hat{U}^2_j\right) - N_\text{res}
$$

Averaging over all indexes $j$ has the advantage that
all possible locations $j$ where systematic errors change sign are taken into account.

Before there are any signs of underfitting, the UFC is negative.
Empirically, we always observe that the UFC slowly decreases with increasing amounts of data,
as long as there is no underfitting.
As soon as the model underfits the data,
the sampling variance of the cumulative sums will quickly exceed
the upper limit of the expectation value and the UFC will increase rapidly.

We observed empirically that the frequency cutoff for which the UFC is minimal
always strikes a good balance between over- and underfitting.
The minimum is always located at the onset of underfitting,
which is barely visible in the residuals to the naked eye.
At this point, one has (nearly) the maximum amount of data in the fit
(and thus minimal uncertainty in the parameters)
without biasing the estimated parameters due to underfitting.


## Akaike's Information Criterion

Akaike's Information Criterion (AIC) is intended for model selection, not data selection,
so, at first glance, it does not seem to be applicable to the cutoff selection.
However, it is possible reformulate data selection as model selection.

The "trick" is to introduce one extra dummy parameter for every discarded data point.
The dummy parameter is simply the model value for that data point and will therefore coincide
with the measurement value.
By introducing these dummy parameters,
all data in the spectrum are always included in the likelihood (irrespective of the cutoff).
The cutoff just determines at which part is modeled with the Exponential Tail model or the dummy
parameters.
Finding the optimal cutoff is therefore reduced to a model selection problem.

In the case of the frequency cutoff,
the dummy parameter corresponds to the scale parameter of the Gamma distribution.
Assuming that $\kappa_k > 1$, the likelihood for that point is maximized by:

$$
    \hat{\theta}_k = \frac{\hat{C}_k}{\kappa_k - 1}.
$$

The log likelihood for such a point becomes:

$$
    \ln \mathcal{L}_k^\text{cut} =
        -\ln\Gamma(\kappa_k)
        - \ln\left(\frac{\hat{C}_k}{\kappa_k - 1}\right)
        + (\kappa_k - 1) \Bigl(\ln(\kappa_k - 1) - 1\Bigr)
$$

To obtain the total log likelihood, $\ln\mathcal{L}^\text{total}$,
the terms $\ln \mathcal{L}_k^\text{cut}$ can be summed over all discarded points
and added to the likelihood of the model fitted to the data before the cutoff.
The AIC with this likelihood then becomes:

$$
    \operatorname{AIC}
    = 2 (N_\text{par} + N_\text{discard}) - 2 \ln\mathcal{L}^\text{total}
$$

where:

- $N_\text{par}$ is the number of parameters in the model for the low-frequency part of the spectrum
  (3 in case of the Exponential Tail model),
- $N_\text{discard}$ is the number of discarded data points


## Cutoff selection

The AIC is the more fundamental of the two criteria and it is also widely used.
Nevertheless, in our numerical benchmarks, the AIC always leads to visually clearly biased fits,
while the UFC systematically selects cutoffs for which the residuals are featureless.

We must emphasize that the UFC is an ad hoc construction.
It is nothing more than a robust mathematical implementation
of how one would intuitively describe underfitting of a smooth model to noisy data.
Nevertheless, its use is well justified by its practical utility,
its simplicity, its effectiveness in our benchmarks,
and the complete absence of manually tunable settings.
