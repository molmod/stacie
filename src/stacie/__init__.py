# Stacie is a STable AutoCorrelation Integral Estimator.
# Copyright (C) 2024-2025 The contributors of the Stacie Python Package.
# See the CONTRIBUTORS.md file in the project root for a full list of contributors.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --
"""The Stacie package."""

from __future__ import annotations

from .estimate import estimate_acint
from .plot import UnitConfig, plot_results
from .spectrum import compute_spectrum
from .summary import summarize_results

__all__ = (
    "UnitConfig",
    "__version__",
    "__version_tuple__",
    "compute_spectrum",
    "estimate_acint",
    "plot_results",
    "summarize_results",
)


try:
    from ._version import __version__, __version_tuple__
except ImportError:
    __version__ = "0.0.0a-dev"
    __version_tuple__ = (0, 0, 0, "a-dev")
