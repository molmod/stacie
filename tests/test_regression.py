# Stacie is a STable AutoCorrelation Integral Estimator.
# Copyright (C) 2024-2025 The contributors of the Stacie Python Package.
# See the CONTRIBUTORS.md file in the project root for a full list of contributors.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --
"""Regression tests for typical Stacie workflows.

Note that the test cases here are not meant as examples of properly analyzed spectra.
(Some of the fits are really poor.)
These test cases explore a few corner cases and verify that Stacie behaves as expected.
"""

from collections.abc import Callable

import pytest
from path import Path

from stacie.cutoff import CV2LCriterion, CV3LCriterion
from stacie.estimate import Result, estimate_acint
from stacie.model import ExpTailModel, PolynomialModel, SpectrumModel
from stacie.msgpack import dump, load
from stacie.plot import plot_results
from stacie.spectrum import Spectrum
from stacie.summary import summarize_results

# Combinations of test spectra and suitable models, with a manual cutoff frequency.
CASES = [
    ("white", PolynomialModel(0), 0.1),
    ("white", PolynomialModel(2), 0.1),
    ("broad", PolynomialModel(0), 0.01),
    ("broad", PolynomialModel(2), 0.1),
    ("pure", ExpTailModel(), 0.02),
    ("pure", PolynomialModel(2, even=True), 0.02),
    ("double", ExpTailModel(), 0.02),
    ("double", PolynomialModel(2, even=True), 0.02),
]
CRITERIA = [CV2LCriterion(), CV3LCriterion(cond=1e6)]


def output_test_result(prefix: str, res: Result | list[Result]):
    """Dump results with Stacie's standard output formats."""
    dn_out = Path("tests/outputs")
    dn_out.makedirs_p()
    plot_results(dn_out / f"{prefix}.pdf", res)
    with open(dn_out / f"{prefix}.txt", "w") as fh:
        fh.write(summarize_results(res))
    path_zip = dn_out / f"{prefix}.nmpk.xz"
    dump(path_zip, res)


def register_result(regtest, res: Result):
    """Register the result of a test with regtest.

    Parameters
    ----------
    res
        The ``stacie.estimate.Result`` to register.
    """
    with regtest:
        print(f"acint = {res.acint:.5e} ± {res.acint_std:.5e}")
        if "corrtime_exp" in res.props:
            print(
                f"corrtime exp = {res.props['corrtime_exp']:.5e} "
                f"± {res.props['corrtime_exp_std']:.5e}"
            )
        print(f"corrtime int = {res.corrtime_int:.5e} ± {res.corrtime_int_std:.5e}")
        print("---")


@pytest.mark.parametrize(("name", "model", "fcut_max"), CASES)
@pytest.mark.parametrize("criterion", CRITERIA)
@pytest.mark.parametrize("full", [True, False])
@pytest.mark.filterwarnings("ignore::stacie.estimate.FCutWarning")
def test_case_scan(
    regtest, name: str, model: SpectrumModel, fcut_max: float, criterion: Callable, full: bool
):
    spectrum = load(f"tests/inputs/spectrum_{name}.nmpk.xz", Spectrum)
    if name == "broad":
        spectrum = spectrum.without_zero_freq()
    result = estimate_acint(spectrum, model, fcut_max=fcut_max, cutoff_criterion=criterion)
    register_result(regtest, result)
    prefix = f"scan_{name}_{model.name}_{criterion.name}"
    output_test_result(prefix, result)


def test_plot_multi():
    cases = [
        ("white", PolynomialModel(2), 0.1),
        ("broad", PolynomialModel(2), 0.1),
    ]
    results = [
        estimate_acint(
            load(f"tests/inputs/spectrum_{name}.nmpk.xz", Spectrum),
            model,
            fcut_max=fcut_max,
            cutoff_criterion=CV2LCriterion(),
        )
        for name, model, fcut_max in cases
    ]
    output_test_result("multi", results)
