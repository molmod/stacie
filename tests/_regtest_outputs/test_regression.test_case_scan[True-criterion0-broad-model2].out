acint = 1.00429e+00 ± 3.77262e-03
corrtime int = 1.70491e+00 ± 6.40447e-03
---
